File: bundle-plugin/src/main/scala/com/yahoo/container/plugin/util/Maps.scala
// Copyright 2016 Yahoo Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.util

import collection.mutable.MultiMap

/**
 * @author  tonytv
 */
object Maps {
  def combine[K, V](map1 : Map[K, V], map2 : Map[K, V])(f : (V, V) => V) : Map[K, V] = {
    def logicError : V = throw new RuntimeException("Logic error.")
    def combineValues(key : K) = key -> f(map1.getOrElse(key, logicError), map2.getOrElse(key, logicError))

    val keysInBoth = map1.keySet intersect map2.keySet
    def notInBoth = !keysInBoth.contains(_ : K)

    map1.filterKeys(notInBoth) ++ map2.filterKeys(notInBoth) ++ keysInBoth.map(combineValues)
  }
}

File: bundle-plugin/src/main/java/com/yahoo/container/plugin/util/Maps.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.util;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.BiFunction;

/**
 * @author Tony Vaagenes
 * @author ollivir
 */
public class Maps {
    public static <K, V> Map<K, V> combine(Map<K, V> left, Map<K, V> right, BiFunction<V, V, V> combiner) {
        Map<K, V> ret = new HashMap<>();
        Set<K> keysRight = new HashSet<>(right.keySet());

        left.forEach((k, v) -> {
            if (keysRight.contains(k)) {
                ret.put(k, combiner.apply(v, right.get(k)));
                keysRight.remove(k);
            } else {
                ret.put(k, v);
            }
        });
        keysRight.forEach(k -> ret.put(k, right.get(k)));

        return ret;
    }
}
